lib/bundler/fetcher.rb

```diff
@@ -374,35 +371,35 @@ def bundler_cert_store
       store
     end
 
-    # Attempt to retry with HTTP authentication, if it's appropriate to do so. Yields to a block;
-    # the caller should use this to re-attempt the failing request with the altered `@remote_uri`.
-    def retry_with_auth
-      # Authentication has already been attempted and failed.
-      raise BadAuthenticationError.new(uri) if @remote_uri.user
-
-      auth = Bundler.settings[@remote_uri.to_s]
-
-      # Authentication isn't provided at all, by "bundle config" or in the URI.
-      raise AuthenticationRequiredError.new(uri) if auth.nil?
+  private
 
-      @remote_uri.user, @remote_uri.password = *auth.split(":", 2)
-      @anonymizable_uri = AnonymizableURI.new(@remote_uri.dup)
-      yield
+    def resolve_remote_uri(uri)
+      add_configured_credentials(Bundler::Source.mirror_for(uri))
     end
 
-  private
+    def add_configured_credentials(uri)
+      auth = Bundler.settings[uri.to_s]
+      if auth
+        uri = uri.dup
+        uri.user, uri.password = *auth.split(":", 2)
```

Comment:
Since ENV and config file are combined, let’s make them lower priority than the in-Gemfile credentials.

> On Jan 8, 2015, at 3:45 PM, Tim Moore <notifications@github.com> wrote:
> 
> In lib/bundler/fetcher.rb <https://github.com/bundler/bundler/pull/3342#discussion-diff-22692558>:
> 
> >      end
> >  
> > -  private
> > +    def add_configured_credentials(uri)
> > +      auth = Bundler.settings[uri.to_s]
> > +      if auth
> > +        uri = uri.dup
> > +        uri.user, uri.password = *auth.split(":", 2)
> Hmm, yeah well right now ENV and config file are normalised to a single settings object.
> 
> —
> Reply to this email directly or view it on GitHub <https://github.com/bundler/bundler/pull/3342/files#r22692558>.
>